### PR TITLE
[PileupJetId, Puppi] Pileup ID input variable fix, puppi weight ValueMap access, optional photon protection for existing puppi weights

### PR DESCRIPTION
**Pull request by jshin96**
PR description:

This PR has three main modifications: 

1) Pileup Jet Id variable bug fix.

PileupJetIdProducer and PileupJetIdAlgo are used to compute input variables and the final discriminant for 
the Pileup jet Id BDT. There was a bug for one of the variable calculation such that when there is no charged 
constituent inside a jet, it assigns the very last constituent in the list of constituent as leading charged 
constituent. This PR corrects this error and assigns zero pt and large phi and eta when there is no charged 
constituent. 

2) Correct puppi weight access implementation by ValueMap.

Previously, the puppi weight for each constituent was accessed from packedCandidate object in the code, but for the 
compatibility with other codes in CMSSW, this weight must be implemented through ValueMap<float>. In this PR, puppi 
weight retrieval from ValueMap is implemented, just like PR #40667. As in that PR, the naming of "puppi weight" is 
generalized to "constituent weight". Furthermore, there are parts of PileupJetIdAlgo where comparison of constituent's pt 
(in order to find, for example, leading charged constituent) was done incorrectly when weighted pt was compared with 
unweighted pt. This error is also fixed.

The ValueMap implementation does not affect CHS jets. However, the fix in leading charged constituent selection could 
have affected CHS jets. In further detail, CHS jets have more constituents in each jet in general and it is very rare 
for CHS jets to have no charged constituent at all, so the effect of such fix was insignificant.

3) Add an option in PuppiProducer to apply photon protection for existing weights.

Previously the photon protection is applied also on existing puppi weights. This will confuse users who would want 
PuppiProducer to provide ValueMap of weights with the same exact weights stored in packedPFCandidates in MiniAODs. 
This PR adds a flag (default set to False) to enable photon protection on existing puppi weights.

PR validation:

- Validation plots comparing the input variables for Pileup Jet Id training before and after fix can be found in this JMAR meeting [contribution](https://indico.cern.ch/event/1245910/#5-pu-jetid-training-for-puppi).
- Passes the usual runTheMatrix test: runTheMatrix.py -l limited -i all --ibeos. Test done by @nurfikri89.
- Passes reMiniAOD and reNanoAOD workflows: runTheMatrix.py -i all --ibeos -l 1325.518,2500.312. Test done by @nurfikri89.
- Passes the JMENano workflows: runTheMatrix.py -i all --ibeos -l 10224.15,11024.15,25202.15,11634.15. Test done by @nurfikri89.

### CODE REVIEW

**Comment by swertz** [comment]
enable nano

**Comment by swertz** [comment]
please test

**Comment by swertz** [comment]
type jme

**Comment by swertz** [comment]
type jetmet

**Comment by swertz** [comment]
Thanks @jshin96 ,

About the bugfix you say:

> In further detail, CHS jets have more constituents in each jet in general and it is very rare for CHS jets to have no charged constituent at all, so the effect of such fix was insignificant.

But looking at the comparison plots, there are rather significant differences for puID for forward jets - which would make sense, since even for CHS these jets wouldn't have many charged consistuents, see eg:

image

Isn't this something we should worry about for current Run2 UL samples? Or would the effect on analysis be negligible because it would be the same in both data and MC?

**Comment by nurfikri89** [comment]
Hi @swertz, 

(Let me answer on behalf of @jshin96 )

> Isn't this something we should worry about for current Run2 UL samples? Or would the effect on analysis be negligible because it would be the same in both data and MC?

I would not worry about it because the trainings were derived with the buggy calculation in place and also when the discriminant is calculated. Also its the same for both data and MC, as you mentioned. In any case, if there is a re-nano (or even a re-mini of Run-2 UL samples), we should definitely re-train. This is the plan anyway when we have to switch to Puppi jets for AK4 jets.

**Comment by swertz** [comment]
+1

Thanks for the clarification @nurfikri89 . @jshin96, can you please prepare a backport to 13_0_X?

**Comment by clacaputo** [comment]
+reconstruction

**Comment by perrotta** [comment]
+1
- Changes in Jet outputs have been approved by the reviewers

File: RecoJets/JetProducers/src/PileupJetIdAlgo.cc [hunk 1]
```diff
@@ -491,7 +522,7 @@ PileupJetIdentifier PileupJetIdAlgo::computeIdVariables(const reco::Jet* jet,
         double dZ0 = 9999.;
         double dZ_tmp = 9999.;
         for (unsigned vtx_i = 0; vtx_i < allvtx.size(); vtx_i++) {
-          const auto& iv = allvtx[vtx_i];
+          auto iv = allvtx[vtx_i];
```
Comment (perrotta) [inline RecoJets/JetProducers/src/PileupJetIdAlgo.cc:525]: Comparing the 12_6 backport with this PR, I notice that this updates was not backported.
While I agree about not backporting it, I wonder why it was even included in the master PR: should it get reverted at some point?

**Pull request merged**